File: 백준/Silver/28279. 덱 2/덱 2.py
import sys

class Deque:
    def __init__(self, capacity):
        self.deque = [0] * capacity
        self.size = 0
        self.head = 0
        self.tail = 0
        self.capacity = capacity

    def add_first(self, value):
        if self.size == 0:
            self.deque[0] = value
            self.head = 0
            self.tail = 0
        else:
            self.head = self.decrease(self.head)
            self.deque[self.head] = value
        self.size += 1

    def add_last(self, value):
        if self.size == 0:
            self.deque[0] = value
            self.head = 0
            self.tail = 0
        else:
            self.tail = self.increase(self.tail)
            self.deque[self.tail] = value
        self.size += 1

    def peek_first(self):
        if self.size > 0:
            return self.deque[self.head]
        return -1

    def peek_last(self):
        if self.size > 0:
            return self.deque[self.tail]
        return -1

    def poll_first(self):
        if self.size > 0:
            value = self.deque[self.head]
            self.head = self.increase(self.head)
            self.size -= 1
            return value
        return -1

    def poll_last(self):
        if self.size > 0:
            value = self.deque[self.tail]
            self.tail = self.decrease(self.tail)
            self.size -= 1
            return value
        return -1

    def get_size(self):
        return self.size

    def increase(self, pointer):
        if pointer == self.capacity - 1:
            return 0
        return pointer + 1

    def decrease(self, pointer):
        if pointer == 0:
            return self.capacity - 1
        return pointer - 1

input = sys.stdin.read
data = input().splitlines()

count = int(data[0])
deque = Deque(count * 2)
output = []

for i in range(1, count + 1):
    tokens = data[i].split()
    command = int(tokens[0])

    if command == 1:
        value = int(tokens[1])
        deque.add_first(value)
    elif command == 2:
        value = int(tokens[1])
        deque.add_last(value)
    elif command == 3:
        output.append(deque.poll_first())
    elif command == 4:
        output.append(deque.poll_last())
    elif command == 5:
        output.append(deque.get_size())
    elif command == 6:
        output.append(1 if deque.get_size() == 0 else 0)
    elif command == 7:
        output.append(deque.peek_first())
    elif command == 8:
        output.append(deque.peek_last())

sys.stdout.write("\n".join(map(str, output)) + "\n")

File: 백준/Silver/28279. 덱 2/덱 2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Deque struct {
	deque    []int
	size     int
	head     int
	tail     int
	capacity int
}

func NewDeque(capacity int) *Deque {
	return &Deque{
		deque:    make([]int, capacity),
		size:     0,
		head:     0,
		tail:     0,
		capacity: capacity,
	}
}

func (d *Deque) addFirst(value int) {
	if d.size == 0 {
		d.deque[0] = value
		d.head = 0
		d.tail = 0
	} else {
		d.head = d.decrease(d.head)
		d.deque[d.head] = value
	}
	d.size++
}

func (d *Deque) addLast(value int) {
	if d.size == 0 {
		d.deque[0] = value
		d.head = 0
		d.tail = 0
	} else {
		d.tail = d.increase(d.tail)
		d.deque[d.tail] = value
	}
	d.size++
}

func (d *Deque) peekFirst() int {
	if d.size > 0 {
		return d.deque[d.head]
	}
	return -1
}

func (d *Deque) peekLast() int {
	if d.size > 0 {
		return d.deque[d.tail]
	}
	return -1
}

func (d *Deque) pollFirst() int {
	if d.size > 0 {
		value := d.deque[d.head]
		d.head = d.increase(d.head)
		d.size--
		return value
	}
	return -1
}

func (d *Deque) pollLast() int {
	if d.size > 0 {
		value := d.deque[d.tail]
		d.tail = d.decrease(d.tail)
		d.size--
		return value
	}
	return -1
}

func (d *Deque) getSize() int {
	return d.size
}

func (d *Deque) increase(pointer int) int {
	if pointer == d.capacity-1 {
		return 0
	}
	return pointer + 1
}

func (d *Deque) decrease(pointer int) int {
	if pointer == 0 {
		return d.capacity - 1
	}
	return pointer - 1
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	line, _ := reader.ReadString('\n')
	count, _ := strconv.Atoi(strings.TrimSpace(line))

	deque := NewDeque(count * 2)

	for i := 0; i < count; i++ {
		line, _ := reader.ReadString('\n')
		tokens := strings.Fields(line)
		command, _ := strconv.Atoi(tokens[0])

		switch command {
		case 1:
			value, _ := strconv.Atoi(tokens[1])
			deque.addFirst(value)
		case 2:
			value, _ := strconv.Atoi(tokens[1])
			deque.addLast(value)
		case 3:
			fmt.Fprintln(writer, deque.pollFirst())
		case 4:
			fmt.Fprintln(writer, deque.pollLast())
		case 5:
			fmt.Fprintln(writer, deque.getSize())
		case 6:
			if deque.getSize() == 0 {
				fmt.Fprintln(writer, 1)
			} else {
				fmt.Fprintln(writer, 0)
			}
		case 7:
			fmt.Fprintln(writer, deque.peekFirst())
		case 8:
			fmt.Fprintln(writer, deque.peekLast())
		}
	}
}
